### ISSUE TITLE
Infinity norm of a vector once more

### ISSUE THREAD

**Issue by rfranke** [OpenModelica/OpenModelica#7640]
This is related to issue #7586. There the assumption was that Modelica would not cover abs(array). However, the function `PowerSystems.Utilities.Math.fminSearch` contains:
```
// Main algorithm
  while (funcount < max_fun and itercount < max_iter) and
    (max(abs(fv[ones(n)] - fv[2:n+1])) > tol_f or
     max(abs(v[:,ones(n)] - v[:,2:n+1])) > tol_x) loop
```
Trying the example `PowerSystems.Examples.AC3ph.Precalculation.Z_matrixTrDat0`, the current code generation appears to manage the first `max(abs(fv[:]))`, but gives up for second `max(abs(v[:,:]))`. It raises the error:
```
[CodegenCFunctions.tpl:7303:14-7303:14:writable] Error: Template error:
   Code generation does not support multiple iterators: array(abs((v[:,fill(1, n)] - v[:,2:n + 1])[$i1, $i2]) for $i2 in 1:n, $i1 in 1:n).
```
Besides not being supported, it also appears inefficient to rebuild the array `(v[:,fill(1, n)] - v[:,2:n + 1])` in each iteration before applying the subscripts `[$i1, $i2]` for only one element of it.

Shouldn't abs(array) just be passed to the code generation, without unrolling it with for loops?

**Comment by perost** [comment]
> Shouldn't abs(array) just be passed to the code generation, without unrolling it with for-loop expressions?

This isn't an issue specifically with `abs` though, but will happen with any function that doesn't take an array with arbitrarily many dimensions (i.e. most of them). I don't like making array constructor expressions either, but it's the only solution that works in all situations (in theory at least, we should obviously make sure that we generate nested array constructors instead if the code generation can't handle multiple iterators, or just fix the code generation).

We can of course decide to implement support for some builtin functions to take arrays with arbitrarily many dimensions. It's not quite as easy as just passing them to the code generation, we also need to make sure they're correctly handled by the backend and the parts of the old frontend that the backend uses (expression simplification, etc.).

**Comment by rfranke** [comment]
It should at least somehow be marked that the iterators don't appear in the expression, so that the code generation can first build the array `(v[:,fill(1, n)] - v[:,2:n + 1])` only once and then apply the iterators afterwards.

**Comment by sjoelund** [comment]
> -- Something like a new ARRAYCALL exp.

It's called an array reduction. We have it already :)
What you would want and what we could also use is a `let`-expression to make some arrays or inline things not be repeated.

Anyway, the code generation does not always create a new array for indexing `v`, right? The Modelica code seems suboptimal at any rate since it calls `max` on something that creates a new array. It would be much better to rewrite the Modelica code to make sure that is guaranteed to not create this temporary array:

```mo
max(abs(v[i,1] - v[i,j]) for i in 1:n, j in 2:n+1)
```

(That is the correct expression, right? A bit hard to see how it should run)

**Comment by rfranke** [comment]
> Anyway, the code generation does not always create a new array for indexing v, right?

I'm afraid it does -- I checked it for the Cpp runtime and the first expression `max(abs(fv[ones(n)] - fv[2:n+1]))`.

> The Modelica code seems suboptimal at any rate since it calls max on something that creates a new array.

I'm not the author of the code, but I know that he liked writing mathematically nice looking expressions and then put a lot of trust in Modelica tools -- maybe not uncommon among Modelica users.

**Comment by perost** [comment]
> > -- Something like a new ARRAYCALL exp.
> 
> It's called an array reduction. We have it already :)
> What you would want and what we could also use is a `let`-expression to make some arrays or inline things not be repeated.
> 
> Anyway, the code generation does not always create a new array for indexing `v`, right? The Modelica code seems suboptimal at any rate since it calls `max` on something that creates a new array. It would be much better to rewrite the Modelica code to make sure that is guaranteed to not create this temporary array:
> 
> ```modelica
> max(abs(v[i,1] - v[i,j]) for i in 1:n, j in 2:n+1)
> ```
> 
> (That is the correct expression, right? A bit hard to see how it should run)

Yeah, I'm not sure either if that's correct, and I'm not sure how easy it would be for the frontend to rewrite expressions like that. An alternative would be to use the form that the old frontend uses:
```Modelica
max(array(abs($i2) for $i2 in $i1) for $i1 in v[:,fill(1, n)] - v[:,2:n + 1])
```
This only works for functions with one vectorized argument though.

**Comment by sjoelund** [comment]
> I'm not the author of the code, but I know that he liked writing mathematically nice looking expressions and then put a lot of trust in Modelica tools -- maybe not uncommon among Modelica users.

Perhaps. But reductions were introduced in order to get good performance out of array operations.

**Comment by sjoelund** [comment]
> Yeah, I'm not sure either if that's correct, and I'm not sure how easy it would be for the frontend to rewrite expressions like that.

We could perhaps in ExpressionSimplify look for `f(array(... for ... in ...))` and rewrite it into `f(... for ... in ...)`. We do support arbitrary expressions for reductions in the code generation...

**Comment by rfranke** [comment]
The REDUCTION's `r.expr` appears to be `abs(v[:,ones(n)] - v[:,2:n+1])($i1,$i2)`. One would need to detect a function call with one argument and then pre-build the code for this argument `v[:,ones(n)] - v[:,2:n+1]` -- assuming that it does not depend on the iterators. I'm afraid one cannot assume this for all reductions!?

**Comment by sjoelund** [comment]
> Edit: How can one identify that we have a function applied to each element of a previously existing array in this case?

If we had let-expressions internally, we wouldn't need to, It could then be:
```mo
let $3 = v[:,ones(n)] - v[:,2:n+1] in
abs($3[$i1,$i2])
```

**Comment by rfranke** [comment]
Looks good to me.
(A sub-question: The translation introduces identifiers starting with `$`. The generated code has become quite hard to read since `System.unquoteIdentifier` replaces `$` with `_omcQ_24`. Is this really needed?)

**Comment by sjoelund** [comment]
> A sub-question: The translation introduces identifiers starting with `$`. The generated code has become quite hard to read since `System.unquoteIdentifier` replaces `$` with `_omcQ_24`. Is this really needed?

Yes. `$` is not a valid identifier in C and some platforms have major issues with it, so we need to replace them in order to handle FMUs. We can't easily choose some other identifier and be guaranteed there are no name clashes in Modelica without it being an invalid C identifier.

**Comment by perost** [comment]
I changed the NF to expand calls with multiple iterators into nested calls in #7642. The question still remains if we can/should implement it in a more efficient way, but this needed to be fixed regardless.

**Comment by rfranke** [comment]
Thank you @perost. This saves a lot of time extending `daeExpReduction` for multiple iterators because: ["The code is quite messy"](https://github.com/OpenModelica/OpenModelica/blob/15713e3c9b8e6f40bea40611529aa1bcb6091ee6/OMCompiler/Compiler/Template/CodegenCFunctions.tpl#L7047).

Would still hope for some let expression to avoid repeated array creations.
